## Lets Take a Look at the [Planetary Systems Dataset](https://exoplanetarchive.ipac.caltech.edu/cgi-bin/TblView/nph-tblView?app=ExoTbls&config=PS) from NASA's Exoplanet Archive

The dataset is messy, I took it in as a csv but there is only one column that contains all the information. Lets get it cleaned up

### Read in the CSV as is

In [40]:
import pandas as pd
df = pd.read_csv('Planetary_Systems.csv', delimiter='\t', skiprows=2)

df.head()

#
0        # COLUMN pl_name:        Planet Name
1   # COLUMN rastr:          RA [sexagesimal]
2           # COLUMN ra:             RA [deg]
3  # COLUMN decstr:         Dec [sexagesimal]
4          # COLUMN dec:            Dec [deg]

#### Notice that there are 71 rows that contain syntax giving all the column names. We will parse through this using regex to create a real dataframe

In [41]:
import re
pattern = r"COLUMN\s+(\w+):"
comments = df['#'].iloc[:71]
comments

0                  # COLUMN pl_name:        Planet Name
1             # COLUMN rastr:          RA [sexagesimal]
2                     # COLUMN ra:             RA [deg]
3            # COLUMN decstr:         Dec [sexagesimal]
4                    # COLUMN dec:            Dec [deg]
                            ...                        
66    # COLUMN sy_kmagerr1:    Ks (2MASS) Magnitude ...
67    # COLUMN sy_kmagerr2:    Ks (2MASS) Magnitude ...
68         # COLUMN rowupdate:      Date of Last Update
69    # COLUMN pl_pubdate:     Planetary Parameter R...
70                # COLUMN releasedate:    Release Date
Name: #, Length: 71, dtype: object

In [42]:
column_names = []
for comment in comments:
    match = re.search(pattern, comment)
    if match:
        column_name = match.group(1)
        column_names.append(column_name)
column_names

['pl_name',
 'rastr',
 'ra',
 'decstr',
 'dec',
 'default_flag',
 'pl_refname',
 'pl_orbper',
 'pl_orbpererr1',
 'pl_orbpererr2',
 'pl_orbperlim',
 'pl_rade',
 'pl_radeerr1',
 'pl_radeerr2',
 'pl_radelim',
 'pl_orbeccen',
 'pl_orbeccenerr1',
 'pl_orbeccenerr2',
 'pl_orbeccenlim',
 'pl_orbincl',
 'pl_orbinclerr1',
 'pl_orbinclerr2',
 'pl_orbincllim',
 'pl_tranmid',
 'pl_tranmiderr1',
 'pl_tranmiderr2',
 'pl_tranmidlim',
 'ttv_flag',
 'pl_imppar',
 'pl_impparerr1',
 'pl_impparerr2',
 'pl_impparlim',
 'pl_trandep',
 'pl_trandeperr1',
 'pl_trandeperr2',
 'pl_trandeplim',
 'pl_trandur',
 'pl_trandurerr1',
 'pl_trandurerr2',
 'pl_trandurlim',
 'pl_ratdor',
 'pl_ratdorerr1',
 'pl_ratdorerr2',
 'pl_ratdorlim',
 'pl_ratror',
 'pl_ratrorerr1',
 'pl_ratrorerr2',
 'pl_ratrorlim',
 'st_refname',
 'st_teff',
 'st_tefferr1',
 'st_tefferr2',
 'st_tefflim',
 'st_rad',
 'st_raderr1',
 'st_raderr2',
 'st_radlim',
 'st_logg',
 'st_loggerr1',
 'st_loggerr2',
 'st_logglim',
 'sy_refname',
 'sy_vmag',
 'sy_v

In [43]:
df = pd.read_csv('Planetary_Systems.csv', delimiter='\t', skiprows=76, names=column_names)
# Display the first few rows of the DataFrame
df.head()

pl_name  rastr  ra  decstr  dec  \
0  55 Cnc e,08h52m35.24s,133.1468373,+28d19m47.34...    NaN NaN     NaN  NaN   
1  55 Cnc e,08h52m35.24s,133.1468373,+28d19m47.34...    NaN NaN     NaN  NaN   
2  55 Cnc e,08h52m35.24s,133.1468373,+28d19m47.34...    NaN NaN     NaN  NaN   
3  55 Cnc e,08h52m35.24s,133.1468373,+28d19m47.34...    NaN NaN     NaN  NaN   
4  55 Cnc e,08h52m35.24s,133.1468373,+28d19m47.34...    NaN NaN     NaN  NaN   

   default_flag  pl_refname  pl_orbper  pl_orbpererr1  pl_orbpererr2  ...  \
0           NaN         NaN        NaN            NaN            NaN  ...   
1           NaN         NaN        NaN            NaN            NaN  ...   
2           NaN         NaN        NaN            NaN            NaN  ...   
3           NaN         NaN        NaN            NaN            NaN  ...   
4           NaN         NaN        NaN            NaN            NaN  ...   

   sy_refname  sy_vmag  sy_vmagerr1  sy_vmagerr2  sy_kmag  sy_kmagerr1  \
0         NaN      NaN          NaN          NaN      NaN          NaN   
1         NaN      NaN          NaN          NaN      NaN          NaN   
2         NaN      NaN          NaN          NaN      NaN          NaN   
3         NaN      NaN          NaN          NaN      NaN          NaN   
4         NaN      NaN          NaN          NaN      NaN          NaN   

   sy_kmagerr2  rowupdate  pl_pubdate  releasedate  
0          NaN        NaN         NaN          NaN  
1          NaN        NaN         NaN          NaN  
2          NaN        NaN         NaN          NaN  
3          NaN        NaN         NaN          NaN  
4          NaN        NaN         NaN          NaN  

[5 rows x 71 columns]

#### Now I see the columns are correct but all the data is in a single column. Luckily they are seperated by columns so lets seperate the info and make  sure it is associated with the correct column

In [44]:
df = df['pl_name'].str.split(',', expand=True)
df.columns = column_names
df.head()

pl_name         rastr           ra         decstr         dec  \
0  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
1  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
2  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
3  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
4  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   

  default_flag                                         pl_refname   pl_orbper  \
0            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...  2.80800000   
1            0  <a refstr=KOKORI_ET_AL__2023 href=https://ui.a...  0.73654625   
2            0  <a refstr=DEMORY_ET_AL__2016 href=https://ui.a...               
3            0  <a refstr=ENDL_ET_AL__2012 href=https://ui.ads...  0.73654600   
4            0  <a refstr=BALUEV_2015 href=https://ui.adsabs.h...  0.73655150   

  pl_orbpererr1 pl_orbpererr2  ...  \
0    0.00200000   -0.00200000  ...   
1    0.00000015   -0.00000015  ...   
2                              ...   
3    0.00000300   -0.00000300  ...   
4    0.00000150   -0.00000150  ...   

                                          sy_refname    sy_vmag sy_vmagerr1  \
0  <a refstr=STASSUN_ET_AL__2019 href=https://ui....  5.9508400   0.0230000   
1  <a refstr=STASSUN_ET_AL__2019 href=https://ui....  5.9508400   0.0230000   
2  <a refstr=STASSUN_ET_AL__2019 href=https://ui....  5.9508400   0.0230000   
3  <a refstr=STASSUN_ET_AL__2019 href=https://ui....  5.9508400   0.0230000   
4  <a refstr=STASSUN_ET_AL__2019 href=https://ui....  5.9508400   0.0230000   

  sy_vmagerr2    sy_kmag sy_kmagerr1 sy_kmagerr2   rowupdate pl_pubdate  \
0  -0.0230000  4.0150000   0.0360000  -0.0360000  2021-02-05    2004-10   
1  -0.0230000  4.0150000   0.0360000  -0.0360000  2024-03-04    2023-03   
2  -0.0230000  4.0150000   0.0360000  -0.0360000  2021-02-05    2016-01   
3  -0.0230000  4.0150000   0.0360000  -0.0360000  2021-02-05    2012-11   
4  -0.0230000  4.0150000   0.0360000  -0.0360000  2021-02-05    2015-01   

  releasedate  
0  2016-10-13  
1  2024-03-04  
2  2016-01-21  
3  2014-05-16  
4  2014-10-29  

[5 rows x 71 columns]

## Project: Grouping ExoPlantes

I'm interested to see if there are noticeable groups of exoplanets given features like:
- **Radius:** The radius of the planet can provide insights into its size and composition. Rocky planets typically have smaller radii compared to gas giants.

- **Mass:** The mass of the planet is another crucial indicator of its composition. Gas giants have much higher masses compared to rocky planets.
    
- **Orbital Period:** The orbital period of the planet, i.e., the time it takes to complete one orbit around its host star, can help in distinguishing between different types of planets. Gas giants tend to have longer orbital periods compared to rocky planets.

- **Transit Depth:** The transit depth, which measures the decrease in brightness of the host star during a planetary transit, can provide information about the size of the planet relative to its host star. Larger transit depths may indicate larger planets.

- **Characteristics of the host star:** Characteristics of the host star, such as its temperature (st_teff), radius (st_rad), and metallicity, can also influence the classification of the planet. For example, the presence of a high-metallicity host star may favor the formation of rocky planets.

- **Orbital Eccentricity:** The eccentricity of the planet's orbit can provide clues about its formation and migration history. Gas giants are more likely to have eccentric orbits compared to rocky planets.

### I have everything I need from this dataset except the mass. But in the Confirmed Planets data from NASA they include planet mass, lets read in the data and put them together

In [34]:
planets_df = pd.read_csv('Confirmed_Planets.csv', delimiter='\t', skiprows=2)

# This df is set up the same as the last so we will use our existing Regex pattern to get column names (92 columns)
comments = planets_df['#'].iloc[:92]

column_names = []
for comment in comments:
    match = re.search(pattern, comment)
    if match:
        column_name = match.group(1)
        column_names.append(column_name)
column_names

['pl_name',
 'hostname',
 'default_flag',
 'sy_snum',
 'sy_pnum',
 'discoverymethod',
 'disc_year',
 'disc_facility',
 'soltype',
 'pl_controv_flag',
 'pl_refname',
 'pl_orbper',
 'pl_orbpererr1',
 'pl_orbpererr2',
 'pl_orbperlim',
 'pl_orbsmax',
 'pl_orbsmaxerr1',
 'pl_orbsmaxerr2',
 'pl_orbsmaxlim',
 'pl_rade',
 'pl_radeerr1',
 'pl_radeerr2',
 'pl_radelim',
 'pl_radj',
 'pl_radjerr1',
 'pl_radjerr2',
 'pl_radjlim',
 'pl_bmasse',
 'pl_bmasseerr1',
 'pl_bmasseerr2',
 'pl_bmasselim',
 'pl_bmassj',
 'pl_bmassjerr1',
 'pl_bmassjerr2',
 'pl_bmassjlim',
 'pl_bmassprov',
 'pl_orbeccen',
 'pl_orbeccenerr1',
 'pl_orbeccenerr2',
 'pl_orbeccenlim',
 'pl_insol',
 'pl_insolerr1',
 'pl_insolerr2',
 'pl_insollim',
 'pl_eqt',
 'pl_eqterr1',
 'pl_eqterr2',
 'pl_eqtlim',
 'ttv_flag',
 'st_refname',
 'st_spectype',
 'st_teff',
 'st_tefferr1',
 'st_tefferr2',
 'st_tefflim',
 'st_rad',
 'st_raderr1',
 'st_raderr2',
 'st_radlim',
 'st_mass',
 'st_masserr1',
 'st_masserr2',
 'st_masslim',
 'st_met',
 'st_me

In [38]:
planets_df = pd.read_csv('Confirmed_Planets.csv', delimiter='\t', skiprows=97, names=column_names)
planets_df = planets_df['pl_name'].str.split(',', expand=True)
planets_df.columns = column_names
planets_df.head()

pl_name hostname default_flag sy_snum sy_pnum  discoverymethod disc_year  \
0  11 Com b   11 Com            0       2       1  Radial Velocity      2007   
1  11 Com b   11 Com            0       2       1  Radial Velocity      2007   
2  11 Com b   11 Com            1       2       1  Radial Velocity      2007   
3  11 UMi b   11 UMi            1       1       1  Radial Velocity      2009   
4  11 UMi b   11 UMi            0       1       1  Radial Velocity      2009   

                            disc_facility              soltype  \
0                        Xinglong Station  Published Confirmed   
1                        Xinglong Station  Published Confirmed   
2                        Xinglong Station  Published Confirmed   
3  Thueringer Landessternwarte Tautenburg  Published Confirmed   
4  Thueringer Landessternwarte Tautenburg  Published Confirmed   

  pl_controv_flag  ... sy_vmagerr2    sy_kmag sy_kmagerr1 sy_kmagerr2  \
0               0  ...  -0.0230000  2.2820000   0.3460000  -0.3460000   
1               0  ...  -0.0230000  2.2820000   0.3460000  -0.3460000   
2               0  ...  -0.0230000  2.2820000   0.3460000  -0.3460000   
3               0  ...  -0.0050000  1.9390000   0.2700000  -0.2700000   
4               0  ...  -0.0050000  1.9390000   0.2700000  -0.2700000   

  sy_gaiamag sy_gaiamagerr1 sy_gaiamagerr2   rowupdate pl_pubdate releasedate  
0  4.4403800      0.0038479     -0.0038479  2014-07-23    2011-08  2014-07-23  
1  4.4403800      0.0038479     -0.0038479  2014-05-14    2008-01  2014-05-14  
2  4.4403800      0.0038479     -0.0038479  2023-09-19    2023-08  2023-09-19  
3  4.5621600      0.0039035     -0.0039035  2018-09-04    2017-03  2018-09-06  
4  4.5621600      0.0039035     -0.0039035  2018-04-25    2011-08  2014-07-23  

[5 rows x 92 columns]

In [45]:
planets_df_subset = planets_df[['pl_name', 'pl_bmasse']]
merged_df = pd.merge(df, planets_df_subset, on='pl_name', how='left')
merged_df.head()

pl_name         rastr           ra         decstr         dec  \
0  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
1  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
2  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
3  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   
4  55 Cnc e  08h52m35.24s  133.1468373  +28d19m47.34s  28.3298154   

  default_flag                                         pl_refname   pl_orbper  \
0            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...  2.80800000   
1            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...  2.80800000   
2            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...  2.80800000   
3            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...  2.80800000   
4            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...  2.80800000   

  pl_orbpererr1 pl_orbpererr2  ...    sy_vmag sy_vmagerr1 sy_vmagerr2  \
0    0.00200000   -0.00200000  ...  5.9508400   0.0230000  -0.0230000   
1    0.00200000   -0.00200000  ...  5.9508400   0.0230000  -0.0230000   
2    0.00200000   -0.00200000  ...  5.9508400   0.0230000  -0.0230000   
3    0.00200000   -0.00200000  ...  5.9508400   0.0230000  -0.0230000   
4    0.00200000   -0.00200000  ...  5.9508400   0.0230000  -0.0230000   

     sy_kmag sy_kmagerr1 sy_kmagerr2   rowupdate pl_pubdate releasedate  \
0  4.0150000   0.0360000  -0.0360000  2021-02-05    2004-10  2016-10-13   
1  4.0150000   0.0360000  -0.0360000  2021-02-05    2004-10  2016-10-13   
2  4.0150000   0.0360000  -0.0360000  2021-02-05    2004-10  2016-10-13   
3  4.0150000   0.0360000  -0.0360000  2021-02-05    2004-10  2016-10-13   
4  4.0150000   0.0360000  -0.0360000  2021-02-05    2004-10  2016-10-13   

  pl_bmasse  
0   8.59000  
1   7.81000  
2            
3            
4   7.46870  

[5 rows x 72 columns]

#### We could reference the [documentation](https://exoplanetarchive.ipac.caltech.edu/applications/DocSet/index.html?doctree=/docs/docmenu.xml&startdoc=1) continuously for what column names mean but I think it would be nice to rename the columns so we can quickly understand what each one means

In [48]:
column_rename_dict = {
    'pl_name': 'planet_name',
    'rastr': 'right_ascension_string',
    'ra': 'right_ascension',
    'decstr': 'declination_string',
    'dec': 'declination',
    'default_flag': 'default_flag',
    'pl_refname': 'planet_reference_name',
    'pl_orbper': 'orbital_period',
    'pl_orbpererr1': 'orbital_period_error_positive',
    'pl_orbpererr2': 'orbital_period_error_negative',
    'pl_orbperlim': 'orbital_period_limit',
    'pl_rade': 'radius_of_planet',
    'pl_radeerr1': 'radius_of_planet_error_positive',
    'pl_radeerr2': 'radius_of_planet_error_negative',
    'pl_radelim': 'radius_of_planet_limit',
    'pl_orbeccen': 'orbital_eccentricity',
    'pl_orbeccenerr1': 'orbital_eccentricity_error_positive',
    'pl_orbeccenerr2': 'orbital_eccentricity_error_negative',
    'pl_orbeccenlim': 'orbital_eccentricity_limit',
    'pl_orbincl': 'orbital_inclination',
    'pl_orbinclerr1': 'orbital_inclination_error_positive',
    'pl_orbinclerr2': 'orbital_inclination_error_negative',
    'pl_orbincllim': 'orbital_inclination_limit',
    'pl_tranmid': 'transit_midpoint',
    'pl_tranmiderr1': 'transit_midpoint_error_positive',
    'pl_tranmiderr2': 'transit_midpoint_error_negative',
    'pl_tranmidlim': 'transit_midpoint_limit',
    'ttv_flag': 'ttv_flag',
    'pl_imppar': 'impact_parameter',
    'pl_impparerr1': 'impact_parameter_error_positive',
    'pl_impparerr2': 'impact_parameter_error_negative',
    'pl_impparlim': 'impact_parameter_limit',
    'pl_trandep': 'transit_depth',
    'pl_trandeperr1': 'transit_depth_error_positive',
    'pl_trandeperr2': 'transit_depth_error_negative',
    'pl_trandeplim': 'transit_depth_limit',
    'pl_trandur': 'transit_duration',
    'pl_trandurerr1': 'transit_duration_error_positive',
    'pl_trandurerr2': 'transit_duration_error_negative',
    'pl_trandurlim': 'transit_duration_limit',
    'pl_ratdor': 'ratio_of_distance_to_star',
    'pl_ratdorerr1': 'ratio_of_distance_to_star_error_positive',
    'pl_ratdorerr2': 'ratio_of_distance_to_star_error_negative',
    'pl_ratdorlim': 'ratio_of_distance_to_star_limit',
    'pl_ratror': 'ratio_of_radii',
    'pl_ratrorerr1': 'ratio_of_radii_error_positive',
    'pl_ratrorerr2': 'ratio_of_radii_error_negative',
    'pl_ratrorlim': 'ratio_of_radii_limit',
    'st_refname': 'star_reference_name',
    'st_teff': 'effective_temperature_of_star',
    'st_tefferr1': 'effective_temperature_of_star_error_positive',
    'st_tefferr2': 'effective_temperature_of_star_error_negative',
    'st_tefflim': 'effective_temperature_of_star_limit',
    'st_rad': 'radius_of_star',
    'st_raderr1': 'radius_of_star_error_positive',
    'st_raderr2': 'radius_of_star_error_negative',
    'st_radlim': 'radius_of_star_limit',
    'st_logg': 'surface_gravity_of_star',
    'st_loggerr1': 'surface_gravity_of_star_error_positive',
    'st_loggerr2': 'surface_gravity_of_star_error_negative',
    'st_logglim': 'surface_gravity_of_star_limit',
    'sy_refname': 'system_reference_name',
    'sy_vmag': 'visual_magnitude_of_star',
    'sy_vmagerr1': 'visual_magnitude_of_star_error_positive',
    'sy_vmagerr2': 'visual_magnitude_of_star_error_negative',
    'sy_kmag': 'infrared_magnitude_of_star',
    'sy_kmagerr1': 'infrared_magnitude_of_star_error_positive',
    'sy_kmagerr2': 'infrared_magnitude_of_star_error_negative',
    'rowupdate': 'row_update_date',
    'pl_pubdate': 'planet_publication_date',
    'releasedate': 'release_date',
    'pl_bmasse': 'planet_mass'
    }

In [49]:
df_renamed = merged_df.rename(columns=column_rename_dict)
df_renamed.head()

planet_name right_ascension_string right_ascension declination_string  \
0    55 Cnc e           08h52m35.24s     133.1468373      +28d19m47.34s   
1    55 Cnc e           08h52m35.24s     133.1468373      +28d19m47.34s   
2    55 Cnc e           08h52m35.24s     133.1468373      +28d19m47.34s   
3    55 Cnc e           08h52m35.24s     133.1468373      +28d19m47.34s   
4    55 Cnc e           08h52m35.24s     133.1468373      +28d19m47.34s   

  declination default_flag                              planet_reference_name  \
0  28.3298154            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...   
1  28.3298154            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...   
2  28.3298154            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...   
3  28.3298154            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...   
4  28.3298154            0  <a refstr=MCARTHUR_ET_AL__2004 href=https://ui...   

  orbital_period orbital_period_error_positive orbital_period_error_negative  \
0     2.80800000                    0.00200000                   -0.00200000   
1     2.80800000                    0.00200000                   -0.00200000   
2     2.80800000                    0.00200000                   -0.00200000   
3     2.80800000                    0.00200000                   -0.00200000   
4     2.80800000                    0.00200000                   -0.00200000   

   ... visual_magnitude_of_star visual_magnitude_of_star_error_positive  \
0  ...                5.9508400                               0.0230000   
1  ...                5.9508400                               0.0230000   
2  ...                5.9508400                               0.0230000   
3  ...                5.9508400                               0.0230000   
4  ...                5.9508400                               0.0230000   

  visual_magnitude_of_star_error_negative infrared_magnitude_of_star  \
0                              -0.0230000                  4.0150000   
1                              -0.0230000                  4.0150000   
2                              -0.0230000                  4.0150000   
3                              -0.0230000                  4.0150000   
4                              -0.0230000                  4.0150000   

  infrared_magnitude_of_star_error_positive  \
0                                 0.0360000   
1                                 0.0360000   
2                                 0.0360000   
3                                 0.0360000   
4                                 0.0360000   

  infrared_magnitude_of_star_error_negative row_update_date  \
0                                -0.0360000      2021-02-05   
1                                -0.0360000      2021-02-05   
2                                -0.0360000      2021-02-05   
3                                -0.0360000      2021-02-05   
4                                -0.0360000      2021-02-05   

  planet_publication_date release_date planet_mass  
0                 2004-10   2016-10-13     8.59000  
1                 2004-10   2016-10-13     7.81000  
2                 2004-10   2016-10-13              
3                 2004-10   2016-10-13              
4                 2004-10   2016-10-13     7.46870  

[5 rows x 72 columns]

#### Lets save this new DF so that we can maybe use it for another project later

In [52]:
df_renamed.to_csv('My_Planet_Features.csv', index=False)

## Create a Dataframe with key features needed to identify a planet

In [56]:
desired_columns = [
    'planet_name',
    'radius_of_planet',
    'planet_mass',
    'orbital_period',
    'transit_depth',
    'orbital_eccentricity',
    'effective_temperature_of_star',
    'radius_of_star',
    'orbital_inclination',
    'impact_parameter',
    'transit_duration',
    'ratio_of_distance_to_star',
    'ratio_of_radii'
]
exoplanet_class_df = df_renamed[desired_columns]

In [57]:
exoplanet_class_df.head()

planet_name radius_of_planet planet_mass orbital_period transit_depth  \
0    55 Cnc e                      8.59000     2.80800000                 
1    55 Cnc e                      7.81000     2.80800000                 
2    55 Cnc e                                  2.80800000                 
3    55 Cnc e                                  2.80800000                 
4    55 Cnc e                      7.46870     2.80800000                 

  orbital_eccentricity effective_temperature_of_star radius_of_star  \
0             0.174000                                                
1             0.174000                                                
2             0.174000                                                
3             0.174000                                                
4             0.174000                                                

  orbital_inclination impact_parameter transit_duration  \
0                                                         
1                                                         
2                                                         
3                                                         
4                                                         

  ratio_of_distance_to_star ratio_of_radii  
0                                           
1                                           
2                                           
3                                           
4

#### There appear to be many null values, lets get a count and clean

In [59]:
exoplanet_class_df.isnull().sum()

planet_name                      0
radius_of_planet                 0
planet_mass                      0
orbital_period                   0
transit_depth                    0
orbital_eccentricity             0
effective_temperature_of_star    0
radius_of_star                   0
orbital_inclination              0
impact_parameter                 0
transit_duration                 0
ratio_of_distance_to_star        0
ratio_of_radii                   0
dtype: int64

It appears the null values we are observing are some kind of empty space, lets replace that with null then observe

In [60]:
# Replace any string that consists only of whitespace with NaN
exoplanet_class_df = exoplanet_class_df.applymap(lambda x: np.nan if isinstance(x, str) and x.strip() == '' else x)

# Now check for null values again
null_counts = exoplanet_class_df.isnull().sum()
print(null_counts)

/var/folders/tk/g8r2_d7531981gcgw9d39_n80000gn/T/ipykernel_65294/2832905327.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  exoplanet_class_df = exoplanet_class_df.applymap(lambda x: np.nan if isinstance(x, str) and x.strip() == '' else x)


planet_name                           0
radius_of_planet                  81936
planet_mass                      286762
orbital_period                    25084
transit_depth                    146945
orbital_eccentricity             161621
effective_temperature_of_star     15114
radius_of_star                    11440
orbital_inclination              157561
impact_parameter                 157143
transit_duration                 111069
ratio_of_distance_to_star        158864
ratio_of_radii                   131887
dtype: int64


There are alot of null values! Mass is going to be a key feature for identifying planet types so instead of aggregating lets drop all null values. Once we have a good grasp on what features determine the type of planet we could create a model that counts key identifiers per classification and come back to label planets with null values for likelihood of a certain class

In [67]:
nulls_dropped_df = exoplanet_class_df.dropna().copy()
nulls_dropped_df

planet_name radius_of_planet planet_mass orbital_period transit_depth  \
108       55 Cnc e            2.080     8.59000     0.73654370      0.041000   
109       55 Cnc e            2.080     7.81000     0.73654370      0.041000   
112       55 Cnc e            2.080     7.46870     0.73654370      0.041000   
113       55 Cnc e            2.080     8.08000     0.73654370      0.041000   
114       55 Cnc e            2.080     8.37000     0.73654370      0.041000   
...            ...              ...         ...            ...           ...   
312252      XO-7 b           15.390   225.34147     2.86414240      0.909000   
312266    pi Men c            2.042     4.82000     6.26790000      0.030000   
312267    pi Men c            2.042     3.49611     6.26790000      0.030000   
312269    pi Men c            2.042     4.52000     6.26790000      0.030000   
312271    pi Men c            2.042     3.63000     6.26790000      0.030000   

       orbital_eccentricity effective_temperature_of_star radius_of_star  \
108                0.061000                       5234.00           0.94   
109                0.061000                       5234.00           0.94   
112                0.061000                       5234.00           0.94   
113                0.061000                       5234.00           0.94   
114                0.061000                       5234.00           0.94   
...                     ...                           ...            ...   
312252             0.038000                       6250.00           1.48   
312266             0.000000                       6037.00           1.10   
312267             0.000000                       6037.00           1.10   
312269             0.000000                       6037.00           1.10   
312271             0.000000                       6037.00           1.10   

       orbital_inclination impact_parameter transit_duration  \
108                 87.300           0.1600        1.5960000   
109                 87.300           0.1600        1.5960000   
112                 87.300           0.1600        1.5960000   
113                 87.300           0.1600        1.5960000   
114                 87.300           0.1600        1.5960000   
...                    ...              ...              ...   
312252              83.450           0.7090        2.7720000   
312266              87.456           0.5900        2.9530000   
312267              87.456           0.5900        2.9530000   
312269              87.456           0.5900        2.9530000   
312271              87.456           0.5900        2.9530000   

       ratio_of_distance_to_star ratio_of_radii  
108                     3.517000        0.02020  
109                     3.517000        0.02020  
112                     3.517000        0.02020  
113                     3.517000        0.02020  
114                     3.517000        0.02020  
...                          ...            ...  
312252                  6.430000        0.09532  
312266                 13.380000        0.01703  
312267                 13.380000        0.01703  
312269                 13.380000        0.01703  
312271                 13.380000        0.01703  

[4491 rows x 13 columns]

In [68]:
len(nulls_dropped_df)

4491

We are left with 4491 rows, but now I see there are multiple rows with the same planet name but differing measurements

#### Create a Dataframe that is 1 row per planet that aggregates the mean of all the measurements gathered for that planet

In [77]:
nulls_dropped_df.dtypes

planet_name                       object
radius_of_planet                 float64
planet_mass                      float64
orbital_period                   float64
transit_depth                    float64
orbital_eccentricity             float64
effective_temperature_of_star    float64
radius_of_star                   float64
orbital_inclination              float64
impact_parameter                 float64
transit_duration                 float64
ratio_of_distance_to_star        float64
ratio_of_radii                   float64
dtype: object

In [76]:
numerical_columns = nulls_dropped_df.iloc[:, 1:].columns
for column in numerical_columns:
    nulls_dropped_df[column] = nulls_dropped_df[column].astype('float')

In [93]:
# Perform the groupby and mean operation on all relevant columns
means = nulls_dropped_df.groupby('planet_name').mean()

# Reset the index to get 'planet_name' as a column
means_df = means.reset_index()

# Display the resulting DataFrame
means_df.head()

planet_name  radius_of_planet  planet_mass  orbital_period  transit_depth  \
0    55 Cnc e             2.080     8.667005        0.736544        0.04100   
1    AU Mic b             3.957    22.221567        8.463080        0.23790   
2    AU Mic c             2.522    14.900000       18.859690        0.09700   
3  CoRoT-10 b            10.870   875.081100       13.240600        1.61036   
4  CoRoT-11 b            16.030   777.085875        2.994330        1.14490   

   orbital_eccentricity  effective_temperature_of_star  radius_of_star  \
0               0.06100                         5234.0            0.94   
1               0.00577                         3678.0            0.74   
2               0.00338                         3678.0            0.74   
3               0.53000                         5075.0            0.79   
4               0.00000                         6440.0            1.37   

   orbital_inclination  impact_parameter  transit_duration  \
0               87.300             0.160            1.5960   
1               89.579             0.134            3.4927   
2               89.227             0.300            4.2360   
3               88.550             0.850            2.9800   
4               83.170             0.818            2.5009   

   ratio_of_distance_to_star  ratio_of_radii  
0                      3.517          0.0202  
1                     18.790          0.0488  
2                     32.050          0.0311  
3                     31.330          0.1269  
4                      6.890          0.1070

#### Finally we have a dataframe we can use to group planets together based on their key features. 

Lets save the Dataframe and jump to our modeling notebook where we will perform Clustering to create clear relationships between planets

In [97]:
means_df.to_csv('planet_clustering.csv', index=False)